patches/net/minecraft/client/renderer/block/model/BlockModelDefinition.java.patch

```diff
@@ -0,0 +1,36 @@
+--- a/net/minecraft/client/renderer/block/model/BlockModelDefinition.java
++++ b/net/minecraft/client/renderer/block/model/BlockModelDefinition.java
+@@ -28,9 +_,10 @@
+ @OnlyIn(Dist.CLIENT)
+ public record BlockModelDefinition(
+     Optional<BlockModelDefinition.SimpleModelSelectors> simpleModels, Optional<BlockModelDefinition.MultiPartDefinition> multiPart
++    , Optional<net.neoforged.neoforge.client.model.block.CustomBlockModelDefinition> customDefinition
+ ) {
+     static final Logger LOGGER = LogUtils.getLogger();
+-    public static final Codec<BlockModelDefinition> CODEC = RecordCodecBuilder.<BlockModelDefinition>create(
++    public static final com.mojang.serialization.MapCodec<BlockModelDefinition> VANILLA_CODEC = RecordCodecBuilder.<BlockModelDefinition>mapCodec(
+             p_409074_ -> p_409074_.group(
+                     BlockModelDefinition.SimpleModelSelectors.CODEC.optionalFieldOf("variants").forGetter(BlockModelDefinition::simpleModels),
+                     BlockModelDefinition.MultiPartDefinition.CODEC.optionalFieldOf("multipart").forGetter(BlockModelDefinition::multiPart)
+@@ -42,8 +_,21 @@
+                 ? DataResult.error(() -> "Neither 'variants' nor 'multipart' found")
+                 : DataResult.success(p_409078_)
+         );
++    public static final Codec<BlockModelDefinition> CODEC = net.neoforged.neoforge.client.model.block.BlockStateModelHooks.makeDefinitionCodec();
++
++    public BlockModelDefinition(Optional<BlockModelDefinition.SimpleModelSelectors> simpleModels, Optional<BlockModelDefinition.MultiPartDefinition> multiPart) {
```

Comment:
Might be nice to also add a `net.neoforged.neoforge.client.model.block.CustomBlockModelDefinition` constructor for datagen.